### ISSUE TITLE
Handling patches CVEs

### ISSUE THREAD

**Issue by AMDmi3** [repology/repology-updater#1045]
Sometimes when software is affected by vulnerability there's no new release in which it is fixed (yet), so distributions have to patch it instead of waiting for a new release. Also there are stable distributions which stick to older versions and patch them for CVEs and bugs instead of updating to a newer versions. In both cases there needs to be a way for distributions to convey that they are unaffected by specific CVEs.

Here are examples of existing tools:
- [Debian security](https://www.debian.org/security/)
- [Gentoo security database](https://glsa.gentoo.org/)
- [FreeBSD vuxml](https://svn.freebsd.org/ports/head/security/vuxml/vuln.xml)

Unfortunately they all share the same property of relying on distro-specific version comparison algorithm in order to determine whether specific package version is vulnerable and in some cases lacking a machine readable form. For instance:

- [Debian](https://www.debian.org/security/2020/dsa-4684):
  > For the stable distribution (buster), this problem has been fixed in version 3.27-6+deb10u1.

- [Gentoo](https://glsa.gentoo.org/glsa/202005-02):
  > Unaffected versions | >= 4.2.0-r5

- FreeBSD
  ```xml
  <vuln vid="88760f4d-8ef7-11ea-a66d-4b2ef158be83">
    <topic>mailman -- arbitrary content injection vulnerability via options or private archive login pages</topic>
    <affects>
      <package>
        <name>mailman</name>
        <range><lt>2.1.30_4</lt></range>
        <range><ge>2.1.31</ge><lt>2.1.33</lt></range>
      </package>
  ```

In all cases specific version algorithms are needed to be implemented in order to match these entries against packages to determine which of them are not affected by mentioned vulnerabilities. Repology supports 70+ repository families and it would be too huge task to implement version comparison algorithms for all of them. Additionally, we would like to encourage community to come with a common machine readable format of publishing information on vulnerability status to allow other (probably more security centric than Repology) tools to be introduced, without the need to reimplement this huge set of version comparison and parsing algorithms.

So, although I absolutely expect complaints from (especially stable) distros on false positive CVE linkages to their (patched) package, there's no way to fix it without their cooperation and I encourage a discussion.

**Comment by AMDmi3** [comment]
Some quick thoughts on the desired properties of a format usable for distro-agnostic information on patched vulnerabilities sharing:
- Commonly used machine-readable format as a base (JSON)
- Which describes a list of facts...
- ...of mitigating [a set of] vulnerabilities in [a set of] packages, with a reasonable amount of additional details
- Notably, package identifiers should be informative and flexible enough to precisely identify packages in an environment where multiple package types (source; binary for different architectures), multiple repositories and repository branches exist, not residing to a version comparison. It should also take into account that a vulnerable package and a fixed update for it may exist in different repositories (probably Debian case).
- Preferably, the format should (mandatory) reference a complete definition of a mitigation (e.g a patch), so the mitigation correctness could be audited and/or ported into another environment

Unfortunately it looks like the only way to precisely identify packages without version comparisons would be to list ALL exact names and versions verbatim which would require much work on distributions side (e.g. a distroy needs to track all package versions/revisions potentially affected by the vulnerability and list all of them), which may compare to implementing a lot of version comparison algs a considerable alternative.

Anyway, some examples somewhat modelled after the above mentioned cases.

```json
{                                                                                                                                                                                                                                                                                                            
  "version": "1.0",
  "mitigations": [
    {
      "vulnerabilities": ["CVE-2020-11102", "CVE-2020-1711", "CVE-2020-7039"],
      "fixes": [
        {
          "type": "patch-tarball",
          "url": "https://dev.gentoo.org/~tamiko/distfiles/qemu-4.2.0-patches-r2.tar.xz"
        }
      ],
      "packages": [
        {
          "names": ["emulation/qemu"],
          "versions": ["4.2.0-r5", "4.2.0-r6", ""4.2.0-r7", "4.2.0-r8", "4.2.0-r9", "4.2.0-r10"]
        }
      ]
    }
  }
}
```

The simple case as we don't need to track binary repository separately. So here's a feed which lists a single fact of mitigation of 3 CVES by patchset in a form of tarball, and it marks the listed versions of package named `emulation/qemu` as fixed. Again note that each next ebuild revision would have to be appended to the list, up to a version not affected by the named CVEs.

---

```json
{
  "version": "1.0",
  "mitigations": [
    {
      "vulnerabilities": ["CVE-2018-13796"],
      "fixes": [
        [
          {
            "type": "patch-raw",
            "url": "https://svn.freebsd.org/ports/head/mail/mailman/files/postfix-verp.diff"
          }
        ]
      ],
      "packages": [
        {
          "source": true,
          "names": ["mail/mailman"],
          "versions: ["2.1.30_4", "2.1.30_5", "2.1.30_6"]
        },
        {
          "source": false,
          "arch": "*",
          "names": ["mailman", "mailman-with-htdig"],
          "versions: ["2.1.30_4", "2.1.30_5", "2.1.30_6"]
        }
      ]
    }
  }
}
```

A bit more interesting FreeBSD case which differentiates source and binary repositories because source and package names differ here.

---

```json
{
  "version": "1.0",
  "mitigations": [
    {
      "vulnerabilities": ["CVE-2020-1763"],
      "fixes": [
        [
          {
            "type": "patch-description",
            "url": "https://salsa.debian.org/debian/libreswan/-/blob/CVE-2020-1763/debian/patches/0006-security-Fix-for-CVE-2020-1763.patch"
          },
          {
            "type": "patch-raw",
            "url": "https://salsa.debian.org/debian/libreswan/-/raw/CVE-2020-1763/debian/patches/0006-security-Fix-for-CVE-2020-1763.patch"
          }
        ]
      ],
      "packages": [
        {
          "repository": "buster",
          "section": "main",
          "source": true,
          "names": ["libreswan"],
          "versions: ["3.27-6+deb10u1", "3.27-6+deb10u2", "3.27-6+deb10u3"],
          "overriding": {
            "versions: ["3.27-6"]
          }
        },
        {
          "repository": "buster",
          "section": "main",
          "source": false,
          "names": ["libreswan", "libreswan-dev", "libreswan-doc"],
          "versions: ["3.27-6+deb10u1", "3.27-6+deb10u2", "3.27-6+deb10u3"],
          "overriding": {
            "versions: ["3.27-6"]
          }
        }
      ]
    }
  }
}
```

A more elaborate case for Debian. First, it lists two alternative forms of a fix - as a web page and as a raw patch. Next, it lists more details on repository location, also differentiates source and binary packages, and additionally lists versions overridden by the named packages - that is, if `libreswan` `3.27-6` and `3.27-6+deb10u1` may coexist in the repository (for example, one comes from _release_ repo and another from _updates_), it is not enough to know that the latter is fixed, but also that the former may be no longer considered as it's overridden by the fixed package for users. Note that this would not be the case for Gentoo where vulnerable and fixed ebuilds may coexist without overriding each other.

**Comment by hmh** [comment]
If you are not aware of security-tracker.debian.org, please take a look.  I believe at least Fedora has something similar.

You will notice you can fetch the Debian security-tracker database at:
https://security-tracker.debian.org/tracker/data/json

Please don't scrap it too often, once a day should be enough.

**Comment by AMDmi3** [comment]
I was aware, but I've totally missed JSON file there, which seems to contain just what I need - facts that a certain vulnerability was fixed for specific package in specific distro version. That is directly usable without the need to implement distro-specific version comparison which takes revisions into account. Of course a specific parser would be needed, but at least now there is a starting point.

**Comment by Anton-Latukha** [comment]
Just wanted to tell that distribution CVEs/patch search is a major feature not present elsewhere.

Hope that you would be able to capitalize on it.

**Comment by jrmarino** [comment]
Is this effort still alive?
I would like to specify which CVEs are patched on a per-port basis on Ravenport's repology.json.
We will use it internally for our own vulnerability audit system, but repology could use it to remove patched projects from the vulnerability list for Ravenports.

**Comment by AMDmi3** [comment]
There was never any effort. After CVE support introduction, the issue was intended as an entry point for ones which may want to complain about patched vulnerabilities still reported. Nobody complained in fact, so I don't plan to work on this.

**Comment by jrmarino** [comment]
Hmm, I wouldn't interpret this as "nobody complained".
You outlined the false positives as shortcoming when the feature was introduced (with an implication that this may change in the future).
I would think the desire for the ability to eliminate false positives is there, but you're seeing acceptance that it's currently not.

I for one would like to see it, but I understand it's unlikely to happen now.
Soon we'll be able to track this on our own.

**Comment by AMDmi3** [comment]
In fact, for repositories which publish a list of patched CVEs along with package information, this would be much easier to implement, as we won't need parallel parser infrastructure (which is not compatible with batch update process anyway), so it's feasible implementing it for Ravenports only, but still it will unlikely happen in python batch updater.

**Comment by jrmarino** [comment]
yes, that's what I meant - each repository provides information on CVEs they've addressed.
repology.json provides the cpe string and in addition it could provide an array listed CVES that have been patched.

e.g.
```
"vulnaddressed": ["CVE-2020-1763",CVE-2018-13410"]
```

something like that.
We'll have that information soon, so it's just a matter of adding to the repology.json file.

**Comment by AMDmi3** [comment]
Sure, feel free to add it, so I have it if/when I get to this feature.
